master/internal/api_runs.go

```diff
@@ -1081,3 +1081,83 @@ func (a *apiServer) PostRunMetadata(
 	}
 	return &apiv1.PostRunMetadataResponse{Metadata: protoutils.ToStruct(result)}, nil
 }
+
+func (a *apiServer) GetRunGroups(ctx context.Context, req *apiv1.GetRunGroupsRequest,
+) (*apiv1.GetRunGroupsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get the user: %s", err)
+	}
+
+	resp := &apiv1.GetRunGroupsResponse{}
+	var groups []*runv1.RunGroup
+	query := db.Bun().NewSelect().
+		Model(&groups).
+		ModelTableExpr("runs AS r").
+		Apply(getRunsGroupsColumns)
+
+	var proj *projectv1.Project
+	if req.ProjectId != nil {
+		proj, err = a.GetProjectByID(ctx, *req.ProjectId, *curUser)
+		if err != nil {
+			return nil, err
+		}
+
+		query = query.Where("r.project_id = ?", req.ProjectId)
+	}
+	if query, err = experiment.AuthZProvider.Get().
+		FilterExperimentsQuery(ctx, *curUser, proj, query,
+			[]rbacv1.PermissionType{rbacv1.PermissionType_PERMISSION_TYPE_VIEW_EXPERIMENT_METADATA},
+		); err != nil {
+		return nil, err
+	}
+	if req.Filter != nil {
+		query, err = filterRunQuery(query, req.Filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if req.Sort != nil {
+		err = sortRuns(req.Sort, query, false)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		query.OrderExpr("group_name ASC")
+	}
+
+	group_name, err := runColumnNameToSQL(req.Group)
+	if err != nil {
+		return nil, err
+	}
+	query.Group(group_name)
+	query.ColumnExpr(fmt.Sprintf("%s AS group_name", group_name))
+	pagination, err := runPagedBunExperimentsQuery(ctx, query, int(req.Offset), int(req.Limit))
+	if err != nil {
+		return nil, err
+	}
+	resp.Pagination = pagination
+	resp.Groups = groups
+	return resp, nil
+}
+
+func getRunsGroupsColumns(q *bun.SelectQuery) *bun.SelectQuery {
+	return q.
+		ColumnExpr("proto_time(min(r.start_time)) AS start_time").
+		ColumnExpr("proto_time(max(r.end_time)) AS end_time").
+		ColumnExpr("avg(r.checkpoint_size) AS checkpoint_size").
+		ColumnExpr("avg(r.checkpoint_count) AS checkpoint_count").
+		ColumnExpr("avg(COALESCE(r.searcher_metric_value)) AS searcher_metric_value").
+		ColumnExpr("extract(epoch FROM coalesce(max(r.end_time), now()) - min(r.start_time))::int AS duration").
+		ColumnExpr("json_agg(distinct e.owner_id) AS user_ids").
+		ColumnExpr("json_agg(distinct e.config->'resources'->>'resource_pool') as resource_pools").
+		ColumnExpr("json_agg(distinct e.config->'searcher'->>'name') as searcher_types").
+		ColumnExpr("json_agg(distinct e.config->'searcher'->>'metric') as searcher_metrics").
+		ColumnExpr("COUNT(*) AS run_count").
+		Join("LEFT JOIN experiments AS e ON r.experiment_id=e.id").
+		Join("LEFT JOIN runs_metadata AS rm ON r.id=rm.run_id").
+		Join("LEFT JOIN users u ON e.owner_id = u.id").
+		Join("LEFT JOIN projects p ON r.project_id = p.id").
+		Join("LEFT JOIN workspaces w ON p.workspace_id = w.id")
```

Comment:
It's beginning to get a little confusing and disjointed, having all these extra table and column aliases off in their own functions just because some other function (`filterRunQuery`) _might_ do something with `rm.metadata`. Not sure what we can do about it just now, but I think it's worth pointing out now and then so we keep thinking about it.

Reply:
We will be returning the metadata for groups in following PRs. For now we are just doing the default columns.